cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/GcpMetadataHttpHandlerInternal.java

```diff
@@ -0,0 +1,196 @@
+              METADATA_FLAVOR_HEADER_KEY, METADATA_FLAVOR_HEADER_VALUE));
+    }
+    responder.sendStatus(HttpResponseStatus.OK,
+        new DefaultHttpHeaders().add(METADATA_FLAVOR_HEADER_KEY, METADATA_FLAVOR_HEADER_VALUE));
+  }
+
+  /**
+   * Returns the token of metadata server.
+   *
+   * @param request The {@link HttpRequest}.
+   * @param responder a {@link HttpResponder} for sending response.
+   * @throws Exception if there is any error.
+   */
+  @GET
+  @Path("/computeMetadata/v1/instance/service-accounts/default/token")
+  public void token(HttpRequest request, HttpResponder responder,
+      @QueryParam("scopes") String scopes) throws Exception {
+
+    LOG.debug("Token requested for namespace: {}", namespace);
+    // check that metadata header is present in the request.
+    if (!request.headers().contains(METADATA_FLAVOR_HEADER_KEY,
+        METADATA_FLAVOR_HEADER_VALUE, true)) {
+      throw new ForbiddenException(
+          String.format("Request is missing required %s header. To access the metadata server, "
+                  + "you must add the %s: %s header to your request.", METADATA_FLAVOR_HEADER_KEY,
+              METADATA_FLAVOR_HEADER_KEY, METADATA_FLAVOR_HEADER_VALUE));
+    }
+
+    // TODO: CDAP-20750
+    if (metadataServiceEndpoint == null) {
+      responder.sendString(HttpResponseStatus.NOT_IMPLEMENTED,
+          String.format("%s has not been set",
+              Constants.TaskWorker.METADATA_SERVICE_END_POINT));
+      return;
+    }
+
+    try {
+      URL url = new URL(metadataServiceEndpoint);
+      if (!Strings.isNullOrEmpty(scopes)) {
+        url = new URL(String.format("%s?scopes=%s", metadataServiceEndpoint, scopes));
+      }
+      io.cdap.common.http.HttpRequest tokenRequest = io.cdap.common.http.HttpRequest.get(url)
+          .addHeader(METADATA_FLAVOR_HEADER_KEY, METADATA_FLAVOR_HEADER_VALUE)
+          .build();
+      HttpResponse tokenResponse = HttpRequests.execute(tokenRequest);
+      responder.sendJson(HttpResponseStatus.OK, tokenResponse.getResponseBodyAsString());
+    } catch (Exception ex) {
+      LOG.warn("Failed to fetch token from metadata service", ex);
+      responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, exceptionToJson(ex),
+          new DefaultHttpHeaders().set(HttpHeaders.CONTENT_TYPE, "application/json"));
+    }
+  }
+
+  /**
+   * Sets the CDAP Namespace information.
+   *
+   * @param request The {@link HttpRequest}.
+   * @param responder a {@link HttpResponder} for sending response.
+   */
+  @PUT
+  @Path("/set-context")
+  public void setContext(FullHttpRequest request, HttpResponder responder)
+      throws BadRequestException {
+    GcpMetadataTaskContext gcpMetadataTaskContext = getGcpMetadataTaskContext(request);
+    this.namespace = new NamespaceId(gcpMetadataTaskContext.getNamespace());
+    SecurityRequestContext.setUserId(gcpMetadataTaskContext.getUserId());
```

Comment:
I don't think this will function properly because SecurityRequestContext stores values in threadlocal variables, and the thread handling this request is not necessarily the same as the thread handling the token request. I'd recommend just directly storing the GcpMetadataTaskContext instead of trying to use SecurityRequestContext.

Reply:
Got it, thanks for the explanation, updated.